### PR TITLE
docs: Clarify correct usage of site PIPs

### LINKED ISSUE #50: Site PIP usage contract
Currently there is an undocumented contract inside nextpnr that site PIPs should only be used for nets that originate/terminate inside that site: https://github.com/YosysHQ/nextpnr/blob/c322cda3f875a5e5dd2575d3a390cbe1cee073e0/fpga_interchange/arch.cc#L1769-L1791

In particular, site PIPs shouldn't be used as general route-throughs, for example through sites that aren't otherwise occupied at all. Tile pseudo-PIPs should be used for these cases.

However, there are two problems:
 - this contract currently isn't documented anywhere (I suspect it might originate from an implicit contract of Vivado) and would be a surprise to anyone implementing another PnR implementation without looking at nextpnr
 - there are edge cases not covered by this as discussed later on in SymbiFlow/python-fpga-interchange#76. In particular for the case where a net re-enters a site via a site pin being used for another purpose. Use of site route-through PIPs should probably be disallowed in this edge case too but we need to formally document the exact allowed/illegal behaviour here and then implement accordingly in nextpnr.

### PR DESCRIPTION
**Pull request by gatecat**
Fixes #50 - hopefully this is an accurate summary of what we intend the contract to be and covers all the possibilities.

### CODE REVIEW

**Review by acomodi — APPROVED** [review]
LGTM

File: docs/device_resources.md [hunk 1]
```diff
@@ -242,6 +242,10 @@ entered via the site port corresponding to the site pin. The first site wire
 in the site will be the site wire attached to the output BEL pin of the site
 port.  From there site routing continues per above.
 
+It is important to note that site PIPs can only be used to access placed cells inside that site. Site PIPs cannot be used as general route-thrus, to route from site input to output. General route-thrus across entire sites should use tile pseudo PIPs as described below - even if a site pin is being validly used for one sink pin of a net that is located inside the site; site PIPs cannot also be used to leave the site again to reach other sinks.
```
Comment (mithro) [inline docs/device_resources.md:245]: Wrapping.

Comment (gatecat) [reply docs/device_resources.md]: Good spot, fixed

File: docs/device_resources.md [hunk 2]
```diff
@@ -242,6 +242,19 @@ entered via the site port corresponding to the site pin. The first site wire
...
+from site input to output. General route-thrus across entire sites should use
+tile pseudo PIPs as described below - even if a site pin is being validly used
+for one sink pin of a net that is located inside the site; site PIPs cannot
+also be used to leave the site again to reach other sinks.
+
+LUT route-thrus, for example, might require both site PIPs and tile pseudo
+PIPs. The site PIP would be used to route through the LUT in order to access
+an associated flipflop input inside the site. The tile PIP would be used to
+route across the entire site as part of the general, inter-tile, routing
+problem.
```
[the remark this answers is not in the document]
Comment (gatecat) [reply docs/device_resources.md:256]: Agreed, happy to do a follow-up with one.

**Pull request merged**